library/std/src/sys/unix/ext/net/ancillary.rs

```diff
@@ -0,0 +1,749 @@
+use crate::convert::TryFrom;
+use crate::io::{self, IoSliceMut};
+use crate::marker::PhantomData;
+use crate::mem::{size_of, zeroed};
+use crate::os::unix::io::RawFd;
+use crate::path::Path;
+use crate::ptr::null_mut;
+use crate::slice::from_raw_parts;
+use crate::sys::unix::ext::net::addr::{sockaddr_un, SocketAddr};
+use crate::sys::unix::net::Socket;
+
+pub(super) fn recv_vectored_with_ancillary_from(
+    socket: &Socket,
+    bufs: &mut [IoSliceMut<'_>],
+    ancillary: &mut SocketAncillary<'_>,
+) -> io::Result<(usize, bool, io::Result<SocketAddr>)> {
+    unsafe {
+        let mut msg_name: libc::sockaddr_un = zeroed();
+
+        let mut msg = libc::msghdr {
+            msg_name: &mut msg_name as *mut _ as *mut _,
+            msg_namelen: size_of::<libc::sockaddr_un>() as libc::socklen_t,
+            msg_iov: bufs.as_mut_ptr().cast(),
+            msg_iovlen: bufs.len(),
+            msg_control: ancillary.buffer.as_mut_ptr().cast(),
+            msg_controllen: ancillary.buffer.len(),
+            msg_flags: 0,
+        };
+
+        let count = socket.recv_msg(&mut msg)?;
+
+        ancillary.length = msg.msg_controllen;
+        ancillary.truncated = msg.msg_flags & libc::MSG_CTRUNC == libc::MSG_CTRUNC;
+
+        let truncated = msg.msg_flags & libc::MSG_TRUNC == libc::MSG_TRUNC;
+        let addr = SocketAddr::from_parts(msg_name, msg.msg_namelen);
+
+        Ok((count, truncated, addr))
+    }
+}
+
+pub(super) fn send_vectored_with_ancillary_to(
+    socket: &Socket,
+    path: Option<&Path>,
+    bufs: &mut [IoSliceMut<'_>],
+    ancillary: &mut SocketAncillary<'_>,
+) -> io::Result<usize> {
+    unsafe {
+        let (mut msg_name, msg_namelen) =
+            if let Some(path) = path { sockaddr_un(path)? } else { (zeroed(), 0) };
+
+        let mut msg = libc::msghdr {
+            msg_name: &mut msg_name as *mut _ as *mut _,
+            msg_namelen,
+            msg_iov: bufs.as_mut_ptr().cast(),
+            msg_iovlen: bufs.len(),
+            msg_control: ancillary.buffer.as_mut_ptr().cast(),
+            msg_controllen: ancillary.length,
+            msg_flags: 0,
+        };
+
+        ancillary.truncated = false;
+
+        socket.send_msg(&mut msg)
+    }
+}
+
+fn add_to_ancillary_data<T: Clone>(
+    buffer: &mut [u8],
+    length: &mut usize,
+    source: &[T],
+    cmsg_level: libc::c_int,
+    cmsg_type: libc::c_int,
+) -> bool {
+    let len = (source.len() * size_of::<T>()) as u32;
+
+    unsafe {
+        let additional_space = libc::CMSG_SPACE(len) as usize;
+        if *length + additional_space > buffer.len() {
+            return false;
+        }
+
+        libc::memset(buffer[*length..].as_mut_ptr().cast(), 0, additional_space);
```

Comment:
Would using `slice::fill` (`slice_fill` feature) be better here? https://github.com/rust-lang/rfcs/issues/2067

Reply:
The `fill` issue is not close yet. Should we wait until is done or should we use it nevertheless?

Reply:
I thought the compiler will try to use nightly features most of the time? I saw usage of unstable features in a lot of place, I think it should be usable. Maybe ask @cuviper?

Reply:
Sure, you can and should use unstable features where they are fitting for implementation. We only need to be careful for unstable features exposed in stable APIs, for example const generics or specialization, but here it's fine.

Even without `slice::fill`, I'd expect using `ptr::write_bytes` rather than reaching out to `libc`.

Reply:
Okay thanks for the information. I have to add the following line to use `fill`: `#![feature(slice_fill)]`. But I do not know in which file the line have to be added.

Reply:
In general `feature` is added to crate root.
In this case it should be `library/std/src/lib.rs`.